File: ios/Orcha/Screens/TaskScreens.swift
import SwiftUI
import UIKit   // UIResponder keyboard notifications (Issue 2 — scroll composer above keyboard)

// Responsibility: Task detail presentation, approval gates, dependencies, thread, and runs.

/* =============================================================================
   Flow 05 — Task detail + thread. Flow 06 — worker runs + streaming log.
   A 1:1 port of the Android `TaskScreens.kt`. These are plain pushed screens:
   the tab's NavigationStack (WorkspaceScreen) owns navigation + destinations.
   ============================================================================= */

/// Flow 05 T4 — task detail: header, flow-08 gate cards, DoD, deps, thread, runs,
/// and the destructive close path (dialog → optional reason alert → cancelTask).
struct TaskDetailScreen: View {
    @Environment(AppModel.self) var model
    @Environment(\.palette) var p
    @Environment(\.dismiss) var dismiss
    let taskId: String

    @State var allRuns = false
    @State var confirmClose = false
    @State var reasonAlert = false
    @State var closeReason = ""
    @State var verifySheetTask: TaskDto?
    @State var planSheetTask: TaskDto?

    var task: TaskDto? { model.snapshot?.tasks.first { $0.id == taskId } }

    var closable: Bool {
        guard let task else { return false }
        return !task.isRoot && task.status != "completed" && task.status != "cancelled"
    }

    var body: some View {
        ScrollView {
            VStack(spacing: 10) {
                if let task {
                    detail(task)
                } else {
                    OrchaCard {
                        Text("Task not found — refresh the workspace.")
                            .foregroundStyle(p.muted)
                    }
                }
                if let error = model.error {
                    Banner(kind: .danger, text: error)
                }
            }
            .padding(16)
        }
        .navigationTitle(task?.title ?? "Task")
        .navigationBarTitleDisplayMode(.inline)
        .toolbar {
            ToolbarItem(placement: .topBarTrailing) {
                Menu {
                    Button("Close task…", role: .destructive) { confirmClose = true }
                        .disabled(!closable)
                } label: {
                    Image(systemName: "ellipsis.circle")
                }
                .accessibilityLabel("Task actions")
            }
        }
        .confirmationDialog(
            "Close \(task?.title ?? "task")?",
            isPresented: $confirmClose,
            titleVisibility: .visible
        ) {
            Button("Close task", role: .destructive) { close(reason: nil) }
            Button("Add reason & close…") { reasonAlert = true }
            Button("Cancel", role: .cancel) {}
        } message: {
            Text("The task is force-closed and anything waiting on it unblocks. A reason is routed to the assignee.")
        }
        .alert("Close task", isPresented: $reasonAlert) {
            TextField("Reason (recommended)", text: $closeReason)
            Button("Close task", role: .destructive) {
                let trimmed = closeReason.trimmingCharacters(in: .whitespacesAndNewlines)
                close(reason: trimmed.isEmpty ? nil : trimmed)
            }
            Button("Cancel", role: .cancel) {}
        } message: {
            Text("The assignee sees this reason on their next wake.")
        }
        .sheet(item: $verifySheetTask) { VerifySheet(task: $0) }
        .sheet(item: $planSheetTask) { PlanApprovalSheet(task: $0) }
        .task { await model.loadTaskDetail(taskId) }
        .refreshable {
            await model.refresh()
            await model.loadTaskDetail(taskId)
        }
    }

    func close(reason: String?) {
        Task {
            if await model.cancelTask(taskId, reason: reason) { dismiss() }
        }
    }

    @ViewBuilder
    func detail(_ task: TaskDto) -> some View {
        headerCard(task)
        gateCards(task)
        descriptionSection(task)
        dodSection(task)
        dependenciesSection(task)
        threadSection
        runsSection
    }

    // MARK: header

    func headerCard(_ task: TaskDto) -> some View {
        OrchaCard {
            HStack(spacing: 8) {
                StatusPill(status: task.status, domain: .task)
                let band = MobileUx.priorityBand(task.priority)
                MetaTag(
                    text: "P\(task.priority ?? 100)",
                    tint: band == .high ? p.danger : band == .elevated ? p.warn : nil
                )
                if task.isRoot { MetaTag(text: "root") }
                Spacer()
            }
            Text(task.title)
                .font(.system(size: 20, weight: .bold))
                .foregroundStyle(p.text)
            HStack(spacing: 8) {
                if let assignee = task.assignees.first ?? task.ownerAlias {
                    AgentAvatar(alias: assignee, size: 30)
                    Text(assignee)
                        .font(.system(size: 13))
                        .foregroundStyle(p.text2)
                } else {
                    Text("unassigned")
                        .font(.system(size: 13))
                        .foregroundStyle(p.faint)
                }
            }
        }
    }

    // MARK: flow-08 violet gate cards

    @ViewBuilder
    func gateCards(_ task: TaskDto) -> some View {
        if task.status == "needs_verification" {
            OrchaCard(borderColor: p.violetLine) {
                Text("AWAITING YOUR VERIFICATION")
                    .font(.system(size: 11, weight: .bold)).tracking(0.8)
                    .foregroundStyle(p.violet)
                Text(task.result ?? "The agent marked this done — review against the definition of done.")
                    .font(.system(size: 13.5))
                    .foregroundStyle(p.text2)
                    .lineLimit(4)
                KitButton(title: "Review & verify", small: true) { verifySheetTask = task }
            }
        }
        if task.planMessage != nil, task.planDecision == nil, task.status == "in_progress" {
            OrchaCard(borderColor: p.violetLine) {
                Text("PLAN AWAITING YOUR APPROVAL")
                    .font(.system(size: 11, weight: .bold)).tracking(0.8)
                    .foregroundStyle(p.violet)
                Text(task.planMessage?.body ?? "")
                    .font(.system(size: 13.5))
                    .foregroundStyle(p.text2)
                    .lineLimit(4)
                KitButton(title: "Review plan", small: true) { planSheetTask = task }
            }
        }
    }

}

File: android/app/src/main/java/io/openorcha/mobile/ui/screens/TaskScreens.kt
package io.openorcha.mobile.ui.screens

/** Owns task detail presentation and its approval and close entry points. */

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.isImeVisible
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.rounded.ArrowBack
import androidx.compose.material.icons.automirrored.rounded.Send
import androidx.compose.material.icons.rounded.ChevronRight
import androidx.compose.material.icons.rounded.MoreVert
import androidx.compose.material.icons.rounded.Refresh
import androidx.compose.material.icons.rounded.Terminal
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.material3.TopAppBarDefaults
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import kotlinx.coroutines.launch
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import io.openorcha.mobile.data.RunDto
import io.openorcha.mobile.data.TaskDto
import io.openorcha.mobile.data.TaskMessageDto
import io.openorcha.mobile.domain.MobileUx
import io.openorcha.mobile.ui.OrchaUiState
import io.openorcha.mobile.ui.components.Avatar
import io.openorcha.mobile.ui.components.AvatarSize
import io.openorcha.mobile.ui.components.Banner
import io.openorcha.mobile.ui.components.BannerKind
import io.openorcha.mobile.ui.components.Bubble
import io.openorcha.mobile.ui.components.BubbleKind
import io.openorcha.mobile.ui.components.DangerTonalButton
import io.openorcha.mobile.ui.components.FeedRow
import io.openorcha.mobile.ui.components.MetaTag
import io.openorcha.mobile.ui.components.OrchaCard
import io.openorcha.mobile.ui.components.OrchaField
import io.openorcha.mobile.ui.components.PrimaryButton
import io.openorcha.mobile.ui.components.SectionH
import io.openorcha.mobile.ui.components.StatusDomain
import io.openorcha.mobile.ui.components.StatusPill
import io.openorcha.mobile.ui.components.TonalButton
import io.openorcha.mobile.ui.theme.MonoSmStyle
import io.openorcha.mobile.ui.theme.MonoStyle
import io.openorcha.mobile.ui.theme.Orcha

/* =============================================================================
   Flow 05 — Task detail + thread. Flow 06 — worker runs + streaming log.
   ============================================================================= */

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun TaskDetailScreen(
    state: OrchaUiState,
    onBack: () -> Unit,
    onRefresh: () -> Unit,
    onOpenThread: () -> Unit,
    onOpenTask: (String) -> Unit,
    onPrepareClose: () -> Unit,
    onCancelTask: (String?) -> Unit,
    onVerify: (Boolean, String?) -> Unit,
    onDecidePlan: (Boolean, String?) -> Unit,
    onOpenRun: (RunDto) -> Unit,
) {
    val p = Orcha.palette
    val task = state.selectedTask
    var menuOpen by remember { mutableStateOf(false) }
    var closing by remember { mutableStateOf(false) }
    var closeReason by remember { mutableStateOf("") }
    var showVerify by remember { mutableStateOf(false) }
    var showPlan by remember { mutableStateOf(false) }
    var allRuns by remember { mutableStateOf(false) }

    Scaffold(
        containerColor = p.bg,
        topBar = {
            TopAppBar(
                colors = TopAppBarDefaults.topAppBarColors(containerColor = Color.Transparent),
                title = { Text(task?.title ?: "Task", maxLines = 1, overflow = TextOverflow.Ellipsis) },
                navigationIcon = { IconButton(onClick = onBack) { Icon(Icons.AutoMirrored.Rounded.ArrowBack, "Back") } },
                actions = {
                    IconButton(onClick = onRefresh) { Icon(Icons.Rounded.Refresh, "Refresh") }
                    val closable = task != null && !task.isRoot && task.status !in setOf("completed", "cancelled")
                    IconButton(onClick = { menuOpen = true }) { Icon(Icons.Rounded.MoreVert, "More") }
                    DropdownMenu(expanded = menuOpen, onDismissRequest = { menuOpen = false }) {
                        DropdownMenuItem(
                            text = { Text("Close task…", color = if (closable) p.danger else p.faint) },
                            enabled = closable,
                            onClick = { menuOpen = false; onPrepareClose(); closing = true },
                        )
                    }
                },
            )
        },
    ) { padding ->
        if (task == null) {
            OrchaCard(Modifier.padding(padding).padding(16.dp)) { Text("Task not found — refresh the workspace.", color = p.muted) }
            return@Scaffold
        }
        LazyColumn(
            modifier = Modifier.fillMaxSize().padding(padding),
            contentPadding = PaddingValues(16.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            item {
                OrchaCard {
                    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                        StatusPill(task.status, StatusDomain.Task)
                        MetaTag("P${task.priority ?: 100}")
                        if (task.isRoot) MetaTag("root")
                    }
                    Text(task.title, style = MaterialTheme.typography.titleLarge)
                    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                        val assignee = task.assignees.firstOrNull() ?: task.ownerAlias
                        if (assignee != null) {
                            Avatar(assignee, human = false, size = AvatarSize.Sm)
                            Text(assignee, style = MaterialTheme.typography.bodyMedium, color = p.text2)
                        } else Text("unassigned", style = MaterialTheme.typography.bodyMedium, color = p.faint)
                    }
                }
            }
            // flow 08 entry points — violet gate cards opening the approval sheets
            if (task.status == "needs_verification") {
                item {
                    OrchaCard(borderColor = p.violetLine) {
                        Text("AWAITING YOUR VERIFICATION", style = MaterialTheme.typography.labelMedium, color = p.violet)
                        Text(task.result ?: "The agent marked this done — review against the definition of done.", color = p.text2, maxLines = 4, overflow = TextOverflow.Ellipsis)
                        PrimaryButton("Review & verify", { showVerify = true }, Modifier.fillMaxWidth(), small = true)
                    }
                }
            }
            if (task.planMessage != null && task.planDecision == null && task.status == "in_progress") {
                item {
                    OrchaCard(borderColor = p.violetLine) {
                        Text("PLAN AWAITING YOUR APPROVAL", style = MaterialTheme.typography.labelMedium, color = p.violet)
                        Text(task.planMessage.body, color = p.text2, maxLines = 4, overflow = TextOverflow.Ellipsis)
                        PrimaryButton("Review plan", { showPlan = true }, Modifier.fillMaxWidth(), small = true)
                    }
                }
            }
            task.description?.takeIf { it.isNotBlank() }?.let {
                item { SectionH("Description") }
                item { OrchaCard { Text(it, color = p.text2) } }
            }
            item { SectionH("Definition of done") }
            item {
                OrchaCard(borderColor = p.accentLine, container = p.surface2) {
                    (task.definitionOfDone ?: "No definition of done was provided.").split("\n").filter { it.isNotBlank() }.forEach {
                        Row(horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                            Text("✓", color = p.accent, fontWeight = FontWeight.W800)
                            Text(it.trim(), color = p.text)
                        }
                    }
                }
            }
            if (task.dependsOn.isNotEmpty()) {
                item { SectionH("Depends on", "${task.dependsOn.size}") }
                items(task.dependsOn, key = { "dep-$it" }) { depId ->
                    val dep = state.snapshot?.tasks?.firstOrNull { it.id == depId }
                    OrchaCard(onClick = { onOpenTask(depId) }) {
                        Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                            Text(if (dep?.status == "completed") "✓" else "🔒", color = if (dep?.status == "completed") p.ok else p.warn)
                            Text(dep?.title ?: depId, style = MaterialTheme.typography.titleSmall, maxLines = 1, overflow = TextOverflow.Ellipsis, modifier = Modifier.weight(1f))
                            dep?.let { StatusPill(it.status, StatusDomain.Task) }
                        }
                    }
                }
            }
            item { SectionH("Thread", "${state.taskMessages.size}") }
            item {
                OrchaCard(onClick = onOpenThread) {
                    Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                        Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(2.dp)) {
                            Text("Thread · ${state.taskMessages.size} messages", style = MaterialTheme.typography.titleSmall)
                            state.taskMessages.lastOrNull()?.let {
                                Text("${it.authorAlias ?: if (it.isHuman) "you" else "agent"}: ${it.body}", style = MaterialTheme.typography.bodyMedium, color = p.muted, maxLines = 1, overflow = TextOverflow.Ellipsis)
                            } ?: Text("No messages yet — say hi.", style = MaterialTheme.typography.bodyMedium, color = p.faint)
                        }
                        Icon(Icons.Rounded.ChevronRight, null, tint = p.faint)
                    }
                }
            }
            item {
                SectionH("Worker runs", "${state.taskRuns.size}", trailing = {
                    if (state.taskRuns.any { it.status == "running" }) StatusPill("running", StatusDomain.Run)
                })
            }
            if (state.taskRuns.isEmpty()) {
                item { OrchaCard { Text("No runs yet — appears when a worker wakes for this task.", color = p.muted) } }
            }
            val visibleRuns = if (allRuns) state.taskRuns else state.taskRuns.take(3)
            items(visibleRuns, key = { it.runId }) { run -> RunRow(run, onOpenRun) }
            if (!allRuns && state.taskRuns.size > 3) {
                item {
                    TextButton(onClick = { allRuns = true }) {
                        Text("All runs (${state.taskRuns.size})", color = p.accent, fontWeight = FontWeight.W700)
                    }
                }
            }
            state.error?.let { item { Banner(BannerKind.Danger, it) } }
        }
    }

    TaskCloseDialog(
        task = task?.takeIf { closing },
        implications = state.closeImplications,
        reason = closeReason,
        onReasonChange = { closeReason = it },
        onDismiss = { closing = false },
        onClose = { closing = false; onCancelTask(closeReason.ifBlank { null }) },
    )
    if (showVerify && task != null) {
        VerifySheet(task, state.actionInFlight, onDismiss = { showVerify = false }) { approve, feedback ->
            showVerify = false; onVerify(approve, feedback)
        }
    }
    if (showPlan && task != null) {
        PlanApprovalSheet(task, state.actionInFlight, onDismiss = { showPlan = false }) { approve, reason ->
            showPlan = false; onDecidePlan(approve, reason)
        }
    }
}
